# BaGFoot
rm( list=objects() )

#install.packages("hash")
#install.packages("digest")
#install.packages("data.table")
#install.packages("Cairo")
#install.packages("aplpack")
#source("https://bioconductor.org/biocLite.R")
#biocLite("BSgenome.Hsapiens.UCSC.hg19")
#biocLite("BSgenome.Mmusculus.UCSC.mm9")

setwd("/lustre/scratch117/cellgen/team204/pm12/BF7")
library(Rsamtools)
library(hash)
library(data.table)
library(digest)
library(Cairo)
library(aplpack)
library(BSgenome.Hsapiens.UCSC.hg19)
library(BSgenome.Mmusculus.UCSC.mm9)
library(BSgenome.Mmusculus.UCSC.mm10)
library(BSgenome.Hsapiens.UCSC.hg38)
library(bagfoot)

source("aux.R")

bamfile1= '/lustre/scratch117/cellgen/team204/pm12/RAWDATA/atac_merged/ATAC11-12.bam';  
cc1 = countReadsBAM(bamfile1, refgenome = "hg19");  # counts the number of cuts in the sequence file.
cc1
cutcountfile1 = makeCutCountBAM(bamfile1, refgenome = "hg38");   # generate a BedGraph file with DNase Cleavages counts

bamfile2= '/lustre/scratch117/cellgen/team204/pm12/RAWDATA/atac_merged/ATAC13-14.bam';
cc2 = countReadsBAM(bamfile2, refgenome = "hg19"); 
cc2
cutcountfile2 = makeCutCountBAM(bamfile2, refgenome = "hg38"); 



#hotspotfile1 = 'S1_Dnmt3aHSC_peaks.csv';   # A peak sites file is comma-separated and must has headers with chromosome, start, end (1-based)
#hotspotfile2 = 'S3_WtHSC_peaks.csv';
#combinedhotspotfile = combineTwoHotspots( hotspotfile1, hotspotfile2, 'S1', 'S3');

#for bedtools multicov
pooledBed <- read.csv('consolidated_hotspot.csv', header=TRUE)
write.table(x=pooledBed[,2:4], file = "pooledBed.bed", append = FALSE, quote = FALSE, sep = "\t", row.names =FALSE, col.names = FALSE)

system(' bedtools multicov -bams /lustre/scratch117/cellgen/team204/pm12/RAWDATA/atac_merged/ATAC11-12.bam -bed pooledBed.bed > NcountsFile1.bed')
system(' bedtools multicov -bams /lustre/scratch117/cellgen/team204/pm12/RAWDATA/atac_merged/ATAC13-14.bam -bed pooledBed.bed > NcountsFile2.bed')
NcountsFile1 <- sum( read.csv('NcountsFile1.bed', head=FALSE , sep="\t")[,4] ,na.rm=TRUE )
NcountsFile2 <- sum( read.csv('NcountsFile2.bed', head=FALSE , sep="\t")[,4] ,na.rm=TRUE )
NcountsFile1
NcountsFile2
rm(pooledBed)
system(' rm pooledBed.bed ' )
system(' rm NcountsFile1.bed ' )
system(' rm NcountsFile2.bed ' )


#tabNoMappability = MakeBiasCorrectionTableBAM(        # This function call generates a hexamer bias frequency table from the given bamfile without mappability assumption
#	bamfile= '/lustre/scratch117/cellgen/team204/pm12/RAWDATA/atac_merged/ATAC17.bam' , #bamfile1,
#	outfile="Hexamer_withoutMap.txt",
#	refgenome="hg38", 
#	np=6, 
#	mapdir='', # if mappability file directory is set to empty, mappability is not used. 
#	atac=TRUE     # Set TRUE for ATAC data. The default value is FALSE
#	);


# run "BaGFoot"

T24 <- CutCount(file = "/lustre/scratch117/cellgen/team204/pm12/BF7/ATAC11-12_AC_cutcount.bgr.gz"  , count = NcountsFile1, name = "T24" ); 
T48 <- CutCount(file = "/lustre/scratch117/cellgen/team204/pm12/BF7/ATAC13-14_AC_cutcount.bgr.gz"  , count = NcountsFile2, name = "T48" );

motifdb_mm10 <- MotifDB(motiflistfile = '/lustre/scratch117/cellgen/team204/pm12/BF/motiflist_hg38_cisbp_filtered.txt',   # This file contains the list of FIMO outputs file
			directory='/lustre/scratch117/cellgen/team204/pm12/BF/hg38_cisbp_fimo/');	                        # Directory of FIMO outputs

gfootoption_fed_fasted <- GFootOption(biasfile = "/lustre/scratch117/cellgen/team204/pm12/BF/Hexamer_withoutMap.txt",   # Hexamer bias frequency table (see "bigfoot_prep_example.R")
					hexamer_pattern= "/lustre/scratch117/cellgen/team204/pm12/BF/nuccode_hg38_6mer_{chr}.dat");		 # nuc. code files generated by MakeBiasCorrectionTableBAM 


# Definition of BiGFoot analysis 
GFoot_fed_fasted= GFoot(control = T24,               ##  Control Data defined above
			treatment = T48, 		 ##  Treatment Data defined above
			sitefile = "/lustre/scratch117/cellgen/team204/pm12/BF7/consolidated_hotspot.csv",  ##  Sites file: Peak File.  This file must have "chr", "st", "ed" in the header.
			motifDB = motifdb_mm10,                 ##  Information about FIMO motifs
			gfootoption= gfootoption_fed_fasted,   ##  Options as defined above
			outputdir = "OUTPUT_T24_vs_T48",  ## OUTPUT directory
			cachedir = "CACHE_T24_vs_T48");   ## CACHE directory 

GFoot_fed_fasted <- run(GFoot_fed_fasted, graphout=T,yrange=c(-2.5,1.5), mc.cores = 8 ) 



dat <- read.table('T24_T48_On_consolidated_hotspot_footprint_depth_table.csv');
gen_bagplot(dat, dataname1='T24', dataname2='T48', factor=2.5);   # generates a bagplot from the calculated footprinting depths

